### PR TITLE
set output.environment.module=true when output.module enabled

### LINKED ISSUE #15292: update default target when `output.module=true`
## Feature request

update defaults when bundling esm library

**What is the expected behavior?**

with
```js
output: {
   module: true
}
```
webpack should set default `output.environment.module=true`

**What is motivation or use case for adding/changing the behavior?**

when building esm bundle/library output environment  should support es modules

**How should this be implemented in your opinion?**

update defaults

**Are you willing to work on this yourself?**
yes

### PR DESCRIPTION
**Pull request by vankop**
**What kind of change does this PR introduce?**

feature 
closes #15292 

**Did you add tests for your changes?**

yes 

**Does this PR introduce a breaking change?**

no, update experimental feature

**What needs to be documented once your changes are merged?**

no

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
Let's add test case to avoid regression

**Comment by vankop** [comment]
@alexander-akait there is no regression, this is update to defaults.. I have updated snapshots

**Review by alexander-akait — APPROVED** [review]

File: lib/config/defaults.js [hunk 1]
```diff
@@ -921,7 +922,11 @@ const applyOutputDefaults = (
 	F(output.environment, "forOf", () => tp && optimistic(tp.forOf));
 	F(output.environment, "bigIntLiteral", () => tp && tp.bigIntLiteral);
 	F(output.environment, "dynamicImport", () => tp && tp.dynamicImport);
-	F(output.environment, "module", () => tp && tp.module);
+	F(
+		output.environment,
+		"module",
+		() => tp && conditionallyOptimistic(tp.module, output.module)
```
Comment (sokra) [inline lib/config/defaults.js:928]:
```suggestion
		() => output.module || (tp && tp.module)
```

File: lib/config/defaults.js [hunk 2]
```diff
@@ -921,7 +922,11 @@ const applyOutputDefaults = (
 	F(output.environment, "forOf", () => tp && optimistic(tp.forOf));
 	F(output.environment, "bigIntLiteral", () => tp && tp.bigIntLiteral);
 	F(output.environment, "dynamicImport", () => tp && tp.dynamicImport);
```
Comment (sokra) [inline lib/config/defaults.js:924]:
```suggestion
	F(output.environment, "dynamicImport", () => output.module || (tp && tp.dynamicImport));
```

**Review by sokra — CHANGES_REQUESTED** [review]
For testing you can remove the `es2020` from some test ESM test cases

File: lib/config/defaults.js (hunk 1 above)
Comment (vankop) [reply lib/config/defaults.js:928]: not sure that this is correct, target could be explicitly not support es modules. e.g. `target: "node 10"`. In my opinion, In this case webpack should throw an error.

File: lib/config/defaults.js [hunk 3]
```diff
@@ -920,8 +921,16 @@ const applyOutputDefaults = (
 	);
 	F(output.environment, "forOf", () => tp && optimistic(tp.forOf));
 	F(output.environment, "bigIntLiteral", () => tp && tp.bigIntLiteral);
-	F(output.environment, "dynamicImport", () => tp && tp.dynamicImport);
-	F(output.environment, "module", () => tp && tp.module);
+	F(
+		output.environment,
+		"dynamicImport",
+		() => tp && conditionallyOptimistic(tp.dynamicImport, output.module)
```
Comment (sokra) [inline lib/config/defaults.js:927]:
```suggestion
		() => conditionallyOptimistic(tp && tp.dynamicImport, output.module)
```

I wonder why it's not that way? Shouldn't be default to `output.module` when there are no target properties?

Comment (vankop) [reply lib/config/defaults.js:927]: yeah, lets do this way.

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Pull request merged**